packages/renderer/MdToHtml/rules/media.ts

```diff
@@ -0,0 +1,46 @@
+import { RuleOptions } from '../../MdToHtml';
+
+const utils = require('../../utils');
+const urlUtils = require('../../urlUtils.js');
+
+function plugin(markdownIt: any, ruleOptions: RuleOptions) {
+	const defaultRender = markdownIt.renderer.rules.link_open;
+
+	markdownIt.renderer.rules.link_open = function(tokens: any, idx: any) {
+		const token = tokens[idx];
+		const href = utils.getAttr(token.attrs, 'href');
+
+		const resourceHrefInfo = urlUtils.parseResourceUrl(href);
+		const isResourceUrl = ruleOptions.resources && !!resourceHrefInfo;
+		if (!isResourceUrl) {
+			return defaultRender(token, idx);
+		}
+
+		const resourceId = resourceHrefInfo.itemId;
+		const result = ruleOptions.resources[resourceId];
+		const resource = result ? result.item : null;
+		const resourceStatus = utils.resourceStatus(ruleOptions.ResourceModel, result);
+
+		if (resourceStatus !== 'ready') {
+			return defaultRender(token, idx);
+		}
+
+		let filename = `./${ruleOptions.ResourceModel.filename(resource)}`;
+		if (ruleOptions.resourceBaseUrl) {
+			filename = ruleOptions.resourceBaseUrl + filename;
+		}
+
+		const mime = resource.mime ? resource.mime.toLowerCase() : '';
+		if (isSupportedType(mime)) {
+			return `<object data="${filename}" style="min-height:100vh;width:100%" type="${mime}"></object> ${defaultRender(tokens, idx)}`;
+		} else {
+			return defaultRender(tokens, idx);
+		}
+	};
+}
```

Comment:
I'm curious how this interacts with the existing link renderer (now in linkReplacement.ts). For example, there's some logic to display a "loading" icon when the resource hasn't been downloaded or decrypted yet - does that still work when this media plugin is enabled? How about the long press on a link, will it work too? And finally there are options like "plainResourceRendering" and "linkRenderingType" which I'm not sure will still work.

Actually, I wonder if all this new logic shouldn't be directly in the existing link_open rule (or in linkReplacement.ts). That will make it easier to make sense of how a link is rendered and will avoid duplicate code. What do you think?

Reply:
> How about the long press on a link, will it work too?
This is not a concern as long press is mobile only and this is not available on mobile.

I'll get back to you with regard to your other comments.

Reply:
> Actually, I wonder if all this new logic shouldn't be directly in the existing link_open rule (or in linkReplacement.ts). That will make it easier to make sense of how a link is rendered and will avoid duplicate code. What do you think?

Won't this make it harder to toggle this functionality via settings?

Reply:
I think none of the approaches is very good - in one case we duplicate complex business logic and make it hard to know how the rules interact with each others (I'm not sure the current implementation fully works, or if it does I don't know how), and in another case, we cram a lot of functionalities in just one rule and make it hard to enable/disable the rule.

I still think the second approach is a little better though, as we can split the code into functions, etc. to make it more readable. As for toggling the functionality we can do that by passing options to the renderer as is done for `checkboxRenderingType` for example. It could be a new options like `enablePdfViewer`.